src/ast/BindingStore.h

```diff
@@ -0,0 +1,103 @@
+/*
+ * Souffle - A Datalog Compiler
+ * Copyright (c) 2020, The Souffle Developers. All rights reserved
+ * Licensed under the Universal Permissive License v 1.0 as shown at:
+ * - https://opensource.org/licenses/UPL
+ * - <souffle root>/licenses/SOUFFLE-UPL.txt
+ */
+
+/************************************************************************
+ *
+ * @file BindingStore.h
+ *
+ * Defines the BindingStore class, which can be used to dynamically
+ * determine the set of bound variables within a given clause.
+ *
+ ***********************************************************************/
+
+#pragma once
+
+#include "souffle/utility/ContainerUtil.h"
+#include <map>
+#include <set>
+#include <string>
+
+namespace souffle {
+
+class AstArgument;
+class AstAtom;
+class AstClause;
+
+class BindingStore {
+public:
+    BindingStore(const AstClause* clause);
+
+    /**
+     * Mark the given variable as strongly bound.
+     * Strongly bound variables can be used to bind functor arguments.
+     * This is the usual case, e.g. body atom appearances
+     */
+    void bindVariableStrongly(std::string varName) {
+        stronglyBoundVariables.insert(varName);
+
+        // Some functor dependencies may be reduced
+        reduceDependencies();
+    }
+
+    /**
+     * Mark the given variable as weakly bound.
+     * Weakly bound variables cannot be used to bind functor arguments.
+     * E.g. bound head arguments in MST adorned relations
+     */
+    void bindVariableWeakly(std::string varName) {
+        weaklyBoundVariables.insert(varName);
+    }
+
+    /** Check if a variable is bound */
+    bool isBound(std::string varName) const {
+        return contains(stronglyBoundVariables, varName) || contains(weaklyBoundVariables, varName);
+    }
+
+    /** Check if an argument is bound */
+    bool isBound(const AstArgument* arg) const;
+
+    /** Counts the number of bound arguments in the given atom */
+    unsigned int numBoundArguments(const AstAtom* atom) const;
```

Comment:
`size_t` would be more consistent than `unsigned int` for the size/count of things.